File: 1. Spark Familiarization/submission/task1.scala
//package homework

import org.json4s.jackson.JsonMethods._
import org.json4s._
import org.json4s.jackson.Serialization.write
import org.json4s.JsonDSL._
import org.apache.spark.sql.SparkSession
import java.io._

object task1 {

  def main(args:Array[String]): Unit = { 
    implicit val formats = org.json4s.DefaultFormats
    val ss = SparkSession
        .builder()
        .appName("scala")
        .config("spark.master", "local[*]")
        .getOrCreate()
    val sc = ss.sparkContext
    val t = sc.textFile(args(0))
    val stopwords = scala.io.Source.fromFile(args(2)).getLines.toList
    val y = args(3)
    val m = args(4).toInt
    val n = args(5).toInt
    var ans = JObject()

    
    ans = ans ~ ("A", t.count())
    ans = ans ~ ("B", t.map(x => parse(x)).filter(x => (x \ "date").values.toString.slice(0,4) == y).count())
    ans = ans ~ ("C", t.map(x => parse(x)).map(x => (x \ "user_id").values.toString).distinct().count())
    var a = t.map(x => parse(x)).map(x => ((x \ "user_id").values.toString, 1)).reduceByKey((x, y) => x + y).sortBy(key => (-key._2, key._1), ascending = true).take(m)
    var k = a.map(t => List(JString(t._1), JInt(t._2))).toList
    ans = ans ~ ("D", k)
    
    a = t.map(x => parse(x)).flatMap(x => (x \ "text").values.toString.split("[ \\(\\[,.!?\n:;\\]\\)]+")).filter(x => !stopwords.contains(x.toLowerCase())).map(x=>(x.toLowerCase().trim(),1)).reduceByKey((x, y) => x + y).sortBy(key => (-key._2, key._1), ascending = true).take(n)
    //k = a.map(t => JString(t)).toList
    ans = ans ~ ("E", a.map(x=>x._1).toList)

    val file = new File(args(1))
    val output = new BufferedWriter(new FileWriter(file))
    output.write(write(ans))
    output.close()
  }
}

File: 1. Spark Familiarization/submission/task1.py
import sys
import json
import re
from pyspark import SparkContext
sc = SparkContext.getOrCreate()


if __name__ == '__main__':
    t = sc.textFile(sys.argv[1])
    output = sys.argv[2]
    stopwords = set()
    with open(sys.argv[3]) as file:
        for i in file:
            stopwords.add(i.split()[0])
    stopwords.add("")
    
    y,m,n = sys.argv[4],sys.argv[5],sys.argv[6]
    ans = dict()
    
    
    ans['A'] = t.count()
    ans['B'] = t.filter(lambda x:json.loads(x)['date'][:4]==y).count()
    ans['C'] = t.map(lambda x:json.loads(x)['user_id']).distinct().count()
    ans['D'] = t.map(lambda x:(json.loads(x)["user_id"],1)).reduceByKey(lambda x,y: x+y).sortBy(lambda x:(-x[1],x[0])).take(int(m))
    ans['E'] = t.map(lambda x:json.loads(x)['text']).flatMap(lambda x:re.split("[ \\(\\[,\n.!?:;\\]\\)]+",x.lower())).filter(lambda x:x.strip() not in stopwords).map(lambda x:(x.strip(),1)).reduceByKey(lambda x,y: x+y).sortBy(lambda x:(-x[1],x[0])).map(lambda x:x[0]).take(int(n))

#.map(lambda x:x[0]).
    with open(output,"w") as file:
        file.write(json.dumps(ans,indent=4))


